Added is_rev field to function definition, Stmt::Expr in statement AST & function reversibility checks

HumbleBee14 opened the pull request:
Introduced an `is_rev` field, and added logic to typecheck that reversible functions are properly annotated and contain only reversible operations. Updated internal helpers to support and validate reversibility constraints

FIXED: FuncType/RevFuncType issues resolved, caused while checking reversibility. Updated other relevant functions which required reversibility checks.

Will add unit tests later to avoid big changes in this PR

HumbleBee14 committed:
Add reversibility checks to function typechecking

Introduces an `is_rev` field and constructor to `FunctionDef`, and adds logic to typecheck that reversible functions are properly annotated and contain only reversible operations. Updates tests and internal helpers to support and validate reversibility constraints.

NOTE: Added TODOs for fixes where currently FuncType/RevFuncType is checked, needs to be fixed with updated FunctionDef

HumbleBee14 committed:
Code refactored

HumbleBee14 committed:
Merge branch 'feature/modulewise-unit-test-structure' into feature/ast-update-is-rev

HumbleBee14 commented:
Note: 'is_rev' and ret_type 'FuncType/RevFuncType' fix discussed earlier is in progress, will be posting in next PR

HumbleBee14 committed:
Merge branch 'rust-ast' into feature/ast-update-is-rev

HumbleBee14 commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -52,19 +52,59 @@ pub fn typecheck_program(prog: &Program) -> Result<(), TypeError> {
-    // Track the expected return type (for return statements).
-    let expected_ret_type = &func.ret_type;
+    let is_annotated_reversible = func.is_rev;
+    let is_signature_reversible = matches!(func.ret_type, Type::RevFuncType { .. }); // TODO: (FIXME) Check RevFuncType or is_reversible?
+
```
This will be removed! Working on overall fix, kept here based on previous understanding (incorrect)

HumbleBee14 committed:
Add Stmt::Expr and improved function type handling for Reversibility

Introduces Stmt::Expr to the AST and typechecker fixes for reversibility.

HumbleBee14 commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -52,19 +52,69 @@ pub fn typecheck_program(prog: &Program) -> Result<(), TypeError> {
+        Type::UnitType
+    } else if func.args.len() == 1 {
+        func.args[0].0.clone()
+    } else {
+        // TODO: For now, if multiple arguments are present and TupleType is not used,
+        // we take the type of the first argument. This needs to be refined
+        // when proper multi-argument function types are introduced (e.g. via TupleType).
+        // TODO: Should fail? Ask Austin
+        func.args[0].0.clone()
```
@ausbin should we consider this case or ignore, if we will never get this case?

ausbin replied:
Man, that is a good question.

I don't want to blow up the scope of this PR, but I think we can add a `TupleType`. I just wrote one here: https://github.com/gt-tinker/qwerty-rust-ast/wiki/Qwerty-AST-grammar#qpu-dsl-grammar. In my opinion, a tuple should have at least two types in it. (That is, it can't get smaller than a pair)

HumbleBee14 replied:
Okay, we will keep this for next PR

HumbleBee14 commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
+            Ok(then_is_rev && else_is_rev)
+        }
+
+        Expr::Variable { name, .. } => {
+            if let Some(var_type) = env.get_var(name) {
+                match var_type {
+                    Type::RevFuncType { .. } => Ok(true),
+                    Type::FuncType { .. } => Ok(false), // Classical functions break reversibility
+                    _ => Ok(true), // Non-function vars don't affect reversibility
+                }
+            } else {
+                Ok(true)
+            }
+        }
```
Is my assumption correct here, except the functions, rest all can be considered reversible if variable?

ausbin replied:
Referencing a variable not in the `env` should be an error, right?

But even if it is in the `env` with a non-reversible type, that doesn't mean the function is inherently irreversible, does it? Consider the following example:
```python
@qpu
@reversible
def z(q: qubit) -> qubit:
    m = {'0','1'}.measure
    return q | {'1'} >> {-'1'}
```

It's silly that `m` is `Assign`ed and never used, but that function should be reversible. This one wouldn't be, though, because `m` is actually called:
```python
@qpu
@reversible
def z2(q: qubit) -> qubit:
    m = {'0','1'}.measure
    unused_bit = '0' | m
    return q | {'1'} >> {-'1'}
```

Do you agree?

HumbleBee14 replied:
Agreed, so I should handle this situation carefully I guess, to handle the scenarios where the non-reversible expressions are specified in the code, but have to check if they are actually used in the code or not. That way, I don't have to worry about the fix here, rather solely rely on the reversibility check method to be reliable in handling these scenarios. 
In fact, the current implementation will throw error anyway, since it'll catch this non-reversible (even if not used) expression used.

HumbleBee14 replied:
Yes, this will throw error and irrespective of that, it's incorrect check here because I'm running **is_expr_inherently_reversible** on expressions before the main typechecking pass **typecheck_stmt** has run. This is a critical yet silly miss from my end. Thanks for pointing this out!

I have changed the order, completing regular checks first before checking for reversibility

HumbleBee14 replied:
Based on our discussion and your fix given in this PR: https://github.com/gt-tinker/qwerty-rust-ast/pull/19/files
I have copied the fixes here

ausbin replied:
Thanks for figuring it out, my apologies for not responding here. I think we're on the same page

ausbin commented on qwerty_ast/src/ast.rs:
```diff
@@ -16,7 +21,7 @@ use crate::dbg::DebugLoc;
     FuncType { in_ty: Box<Type>, out_ty: Box<Type> },
     RevFuncType { in_out_ty: Box<Type> },
-    RegType { elem_ty: RegKind, dim: u64 },
+    RegType { elem_ty: RegKind, dim: u64 }, // TODO: dim: DimExpr instead of u64
```
Don't worry. I'm thinking we'll have a MetaAST data structure with DimExprs that gets expanded into your AST here. I can tell you about it tomorrow

ausbin replied:
There are a couple of other instances of this popping up below, but I don't want to spam you with comments.

I can explain what I'm thinking in the meeting tomorrow

ausbin commented on qwerty_ast/src/ast.rs:
```diff
@@ -5,7 +5,12 @@
+ * ## Version 1.1.0
+ * - Added `is_rev` field to `FunctionDef` for reversible function support
+ * - Added `Stmt::Expr` variant to allow standalone expressions as statements
+ *
```
It's not a bad idea to have a changelog if we end up having third-party users of this, but for now, the git history should suffice. I don't want to demoralize you, though, if you like keeping the changelog

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -160,17 +216,37 @@ pub fn typecheck_expr(expr: &Expr, env: &mut TypeEnv) -> Result<Type, TypeError>
-            Ok(func_ty)
+
+            match func_ty {
+                Type::RevFuncType { in_out_ty } => Ok(Type::RevFuncType { in_out_ty }),
```
I didn't communicate this to you properly, but I am thinking that the operand `e` of `~e` (in terms of your data structure, the `func` of an `Adjoint`) should have an `in_ty` of `qubit[m]`, where $m > 0$. Like this rule:

![Image](https://github.com/user-attachments/assets/a6e803cb-4d9e-4804-8250-edaf94fcd357)

Some of the MLIR work downstream expects this because it made the implementation dramatically easier, so it would be nice if the type system in the frontend could enforce that.

HumbleBee14 replied:
Yes! I remember you had shared this info when I had asked about the Adjoint implementation details. That's a miss on my end, perhaps looking and understanding this maths/quantum jargon sometimes doesn't makes sense to my in one go. 
But is this reversibility check should be generalized in the function "fn **infer_function_body_reversibility** (...)" or not? Or is this specific constraint for stricter qubit[m] (m > 0) check is only needed for constructs like Adjoint, not for all reversible functions? 
I feel instead of adding constraint here, I should rather add this check in the 'infer_function_body_reversibility' method itself, so that this could remain as is.

ausbin replied:
Looks like I didn't reply to this either, sorry for that.

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -160,17 +216,37 @@ pub fn typecheck_expr(expr: &Expr, env: &mut TypeEnv) -> Result<Type, TypeError>
+                    // Classical functions cannot have an adjoint
+                    Err(TypeError {
+                        kind: TypeErrorKind::InvalidType(format!(
+                            "Cannot take adjoint of non-reversible function: {:?}",
```
Thanks for efforts on these error messages. I think it will pay off when we get students to use this

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
+
+        Expr::BasisTranslation { .. } => Ok(true),
+
+        Expr::Predicated {
+            then_func,
+            else_func,
+            ..
+        } => {
+            // Predicated operations can be reversible if both their 'then' and 'else' branches are.
+            let then_is_rev = is_expr_inherently_reversible(then_func, env)?;
+            let else_is_rev = is_expr_inherently_reversible(else_func, env)?;
+            Ok(then_is_rev && else_is_rev)
+        }
```
This code should work fine. But for what it's worth, the type checker should verify that both operands are reversible functions:

![Image](https://github.com/user-attachments/assets/9bb60f67-ffee-4a29-8cdc-a18c6bc94f3e)

That rule says:    
> If `b` is a well-typed $m_2$-qubit basis and    
> if `e_1` is a reversible function operating on $m_1$ qubits and    
> if `e_2` is a reversible function operating on $m_1$ qubits too and    
> if `b` contains `TargetVector` (`'_'`) vectors at exactly $m_1$ positions and    
> if there are some positions in `b` that are neither `TargetVector`s (`'_'`) nor `PadVector`s (`'?'`), then    
> the expression `e_1 if b else e_2` is a reversible function on $m_2$ qubits

ausbin replied:
You don't have to modify `typecheck_expr()` to do this check on `Predicated` nodes in this PR in my opinion, not trying to say that. Just saying.

Also, if it helps, here are two ways to think about why you can predicate only reversible functions:

1. In Qwerty, predicating a function means running it only in a proper subspace. But it doesn't make sense to perform a measurement in a proper subspace of the overall state space. To the best of my knowledge about physics, you cannot measure one qubit only when a second qubit is in the `'1'` state without somehow affecting that second qubit's state.

2. In gate/circuit terms, a Qwerty predication is a generalization of a control (e.g., the control on a CNOT that distinguishes it from a plain NOT). And in quantum circuits, there is no way to write a controlled measurement.

HumbleBee14 replied:
I updated the typechecker to ensure both branches of a predicated expression are reversible functions with matching signatures, as required by the Qwerty type system, in typecheck_expr() method.

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
+                return Ok(false);
+            }
+
+            let rhs_ty = typecheck_expr(rhs, env)?; // Recursively typecheck to get the actual type of RHS
```
Am I correct that this would throw a "variable not found" error for the last line below because `env` is not being populated with `Assign`s as `infer_function_body_reversibility()` iterates through the statements in the body?

```python
@qpu
@reversible
def trolling(q: qubit) -> qubit:
    m = {'0','1'}.measure
    return '0' | m
```

(In AST terms, `@reversible` means `is_rev=true` on the `FunctionDef`; the `=` is an `Assign` statement; and the `m` on the last line is a `Variable` expression)

ausbin replied:
I have an idea below in a comment on `infer_function_body_reversibility()` that could help with this

HumbleBee14 replied:
Yes this will indeed throw error as I'm (was) checking the reversibility before typechecking whole code and inserting all expressions in the type env, which is fixed now by changing the order of check.

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
+/// This function checks all statements and expressions within the body.
+/// It assumes the function's signature (return type) has already been
+/// validated for consistency with the annotation by `typecheck_function`.
+fn infer_function_body_reversibility(
```
I'm wondering if there's a way to integrate this with `typecheck_function()`. For example, maybe `typecheck_function()` could return a `Result<Reversibility, TypeError>` instead of `Result<(), TypeError>`.


I only suggest this `Reversibility` enum is to avoid "Boolean blindness" (if you've heard that term). It could look like this:
```rust
enum Reversiblility {
    NotRev,
    Rev,
}
```

ausbin replied:
This integration idea could help with avoiding duplicate code for e.g. updating the `env`

ausbin replied:
Sorry, I don't know why I said `typecheck_function()`. I should have said `typecheck_stmt` and `typecheck_expr()`

ausbin reviewed (commented):
Here are a few ideas

HumbleBee14 committed:
Temp fix for env variables not present during reversibility checks

HumbleBee14 committed:
Update measurement reversibility logic bug

HumbleBee14 committed:
Refactor reversibility checks in is_expr_inherently_reversible

Simplifies and clarifies the logic for determining if expressions are inherently reversible. Removes special-casing for Measure, Discard, and BasisTranslation, and consolidates control flow and function call handling. Variable reversibility now only checks for variable existence, and most expressions default to reversible unless explicitly handled.

HumbleBee14 commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -52,22 +52,72 @@ pub fn typecheck_program(prog: &Program) -> Result<(), TypeError> {
+            });
+        }
+    }
 
```
Since we allow shadowing, with the updated two pass approach, Austin found the issue mentioned below

<img width="1383" height="328" alt="image" src="https://github.com/user-attachments/assets/3068e48c-d189-4c8f-89bd-8695e5b65b93" />

HumbleBee14 replied:
So basically, the current two-pass approach loses the temporal context of variable bindings and the reversibility checker sees the final state of the environment, not the state at each statement and hencee takes wrong call, makes it reversible meanwhile it should logically break at second line.

To fix this, I am planning to use single pass with statement by statement checking, where each statement sees the environment as it existed at that point.

HumbleBee14 committed:
Refactor function body reversibility checking

Replaces the two-pass reversibility check in typecheck_function with a single-pass approach that checks each statement's reversibility before updating the environment. Removes the infer_function_body_reversibility helper and introduces check_stmt_reversibility for more precise, statement-level checking.

HumbleBee14 committed:
Enforce reversible function types in predicated expressions

Updated type checking for predicated expressions to require both 'then' and 'else' branches to be reversible functions with matching signatures.

HumbleBee14 committed:
Merge branch 'rust-ast' into feature/ast-update-is-rev

HumbleBee14 committed:
Update error.rs

HumbleBee14 committed:
Update typecheck.rs

ausbin commented on qwerty_ast/src/typecheck.rs:
```diff
@@ -941,6 +1086,90 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
+        }
+
+        // Everything else is reversible by default (actual checks for non-reversibity done during execution call)
+        _ => Ok(true),
```
I think we also need `is_expr_inherently_reversible` for the `Tensor` node. (Other than that, I think you have recursively checked all the nested expressions inside expressions, thanks for that — my bad for forgetting the `Tensor` case when we spoke earlier)

ausbin reviewed (approved):
Thanks for your persistence on this. I need some of your code here for the MLIR lowering I was working on, so I am inclined to merge

ausbin commented:
Merged 6433499ab6be87da14ba386ca7b60cd2a3db727f
